src/sentry/incidents/logic.py

```diff
@@ -269,16 +270,28 @@ def delete_comment(activity):
     return activity.delete()
 
 
+def create_pending_incident_snapshot(incident):
+    assert incident.status == IncidentStatus.CLOSED.value
+
+    time_window = incident.alert_rule.time_window if incident.alert_rule is not None else 1
+    target_run_date = incident.current_end_date + min(
+        timedelta(minutes=time_window * 10), timedelta(days=10)
+    )
+    return PendingIncidentSnapshot.objects.create(
+        incident=incident, target_run_date=target_run_date,
+    )
+
+
 def create_incident_snapshot(incident, windowed_stats=False):
     """
     Creates a snapshot of an incident. This includes the count of unique users
     and total events, plus a time series snapshot of the entire incident.
     """
-
     assert incident.status == IncidentStatus.CLOSED.value
 
     event_stats_snapshot = create_event_stat_snapshot(incident, windowed_stats=windowed_stats)
     aggregates = get_incident_aggregates(incident)
+
     return IncidentSnapshot.objects.create(
```

Comment:
Since this is happening asynchronously now, is there any chance we end up attempting to create two snapshots? Like if we resolve, open, resolve an incident in a short timespan, then this will cause there to be two pending snapshots. Then creating the second snapshot will fail due to the `OneToOneConstraint` (which should just be a unique index on the incident column)

Reply:
I've added code to delete the pending snapshot when an incident's status is updated to open (as long as update_incident_status is used) and a check in the task to make sure the incident is still closed and we have no existing snapshot. I think this is redundant, but perhaps better safe than sorry?